Fix S2699 FP: Inspect invoked methods to search for assertions

pavel-mikula-sonarsource opened the pull request:
Fixes #4241

csaba-sagi-sonarsource commented on analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs:
```diff
@@ -98,6 +95,36 @@ protected override void Initialize(SonarAnalysisContext context)
                 SyntaxKind.MethodDeclaration);
         }
 
+        private static bool ContainsAssertion(MethodDeclarationSyntax methodDeclaration, SemanticModel previousSemanticModel, ISet<IMethodSymbol> visitedSymbols, int level)
```
hasAnyAssertion part should be also part of ContainsAssertion. Because this way we do not recognize things like x.Should().Be(y) if it is in a helper method.

csaba-sagi-sonarsource replied:
Also, now that the assert exception part is implemented, it would be nice to rebase on top of master and put hte hasAnyThrownAssertionExceptions  part inside the ContainsAssertion check too.

pavel-mikula-sonarsource replied:
Nice catch! I can simplify the logic and remove redundant `DescendantNodes()` evaluation (that can be expensive).

csaba-sagi-sonarsource commented on analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs:
```diff
@@ -98,6 +95,36 @@ protected override void Initialize(SonarAnalysisContext context)
+                visitedSymbols.Add(symbol);
+                foreach (var invokedDeclaration in symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).OfType<MethodDeclarationSyntax>())
+                {
+                    if (ContainsAssertion(invokedDeclaration, currentSemanticModel, visitedSymbols, level + 1))
```
If I remember correctly we talked about BFS implementation and this one looks like DFS. Or am I missing something?

pavel-mikula-sonarsource replied:
I've skipped the BFS idea to keep it simple, because I've chosen the limit to be 2 instead of the original idea of 3

csaba-sagi-sonarsource reviewed (changes requested):
After addressing the comments, it would be possible to remove the AssertionMethodAttribute Attribute from our unit tests, as we will be able to detect the assertions without the custom attribute.

pavel-mikula-sonarsource commented:
> After addressing the comments, it would be possible to remove the AssertionMethodAttribute Attribute from our unit tests, as we will be able to detect the assertions without the custom attribute.

And after release to SC :)

csaba-sagi-sonarsource reviewed (changes requested):
Can we add new UT cases with  assert exception throwing  and  fluent assertion with Should?

csaba-sagi-sonarsource reviewed (approved):
LGTM
